File: clients/soa_services/message/python/src/crosslab/soa_services/message/message_service.py
import json
from typing import Literal, Union

from crosslab.soa_client.connection import Connection, DataChannel
from crosslab.soa_client.service import Service
from pyee.asyncio import AsyncIOEventEmitter

from crosslab.soa_services.message.messages import (
    MessageServiceConfig,
    MessageServiceEvent,
)


class MessageService__Producer(Service):
    service_type = "https://api.goldi-labs.de/serviceTypes/message"
    service_direction = "producer"
    service_id: str

    def __init__(self, service_id: str):
        self.service_id = service_id

    def getMeta(self):
        return {
            "serviceId": self.service_id,
            "serviceType": self.service_type,
            "serviceDirection": self.service_direction,
        }

    def setupConnection(
        self, connection: Connection, serviceConfig: MessageServiceConfig
    ):
        self.channel = DataChannel()
        if connection.tiebreaker:
            connection.transmit(serviceConfig, "data", self.channel)
        else:
            connection.receive(serviceConfig, "data", self.channel)

    def teardownConnection(self, connection: Connection):
        pass

    async def sendMessage(self, message: str, message_type: Literal["info", "error"]):
        self.channel.send(json.dumps({"messageType": message_type, "message": message}))


class MessageService__Consumer(Service, AsyncIOEventEmitter):
    service_type = "https://api.goldi-labs.de/serviceTypes/message"
    service_direction = "consumer"
    service_id: str

    def __init__(self, service_id: str):
        AsyncIOEventEmitter.__init__(self)
        self.service_id = service_id

    def getMeta(self):
        return {
            "serviceId": self.service_id,
            "serviceType": self.service_type,
            "serviceDirection": self.service_direction,
        }

    def setupConnection(
        self, connection: Connection, serviceConfig: MessageServiceConfig
    ):
        self.channel = DataChannel()
        self.channel.on("data", lambda data: self.handleData(data))
        if connection.tiebreaker:
            connection.transmit(serviceConfig, "data", self.channel)
        else:
            connection.receive(serviceConfig, "data", self.channel)

    def teardownConnection(self, connection: Connection):
        pass

    def handleData(self, data: Union[str, bytes]):
        if isinstance(data, str):
            msg = json.loads(data)
            event: MessageServiceEvent = {
                "message_type": msg["messageType"],
                "message": msg["message"],
            }
            self.emit("message", event)

File: clients/soa_services/message/js/src/message_service.ts
import {
  Consumer,
  Producer,
  Service,
  ServiceConfiguration,
} from '@cross-lab-project/soa-client';
import { DataChannel } from '@cross-lab-project/soa-client';
import { PeerConnection } from '@cross-lab-project/soa-client';
import { TypedEmitter } from 'tiny-typed-emitter';

import { MessageServiceEvent } from './messages';

type ServiceType = 'https://api.goldi-labs.de/serviceTypes/message';
const ServiceType: ServiceType = 'https://api.goldi-labs.de/serviceTypes/message';

export interface MessageServiceConfiguration extends ServiceConfiguration {
  serviceType: ServiceType;
}
function checkConfig(
  config: ServiceConfiguration,
): asserts config is MessageServiceConfiguration {
  if (config.serviceType !== ServiceType) {
    //throw Error("Service Configuration needs to be for MessageService type");
  }
}

export class MessageService__Producer implements Service {
  serviceType = ServiceType;
  serviceId: string;
  serviceDirection = Producer;
  channel?: DataChannel;

  constructor(serviceId: string) {
    this.serviceId = serviceId;
  }

  getMeta() {
    return {
      serviceType: this.serviceType,
      serviceId: this.serviceId,
      serviceDirection: this.serviceDirection,
    };
  }

  setupConnection(connection: PeerConnection, serviceConfig: ServiceConfiguration): void {
    checkConfig(serviceConfig);
    const channel = new DataChannel();
    if (connection.tiebreaker) {
      connection.transmit(serviceConfig, 'data', channel);
    } else {
      connection.receive(serviceConfig, 'data', channel);
    }
    this.channel = channel;
  }

  async sendMessage(message: string, messageType: string) {
    if (this.channel !== undefined) {
      await this.channel.ready();
      this.channel.send(JSON.stringify({ messageType: messageType, message: message }));
    }
  }
}

interface MessageService__ConsumerEvents {
  message: (event: MessageServiceEvent) => void;
}

export class MessageService__Consumer
  extends TypedEmitter<MessageService__ConsumerEvents>
  implements Service
{
  serviceType = ServiceType;
  serviceId: string;
  serviceDirection = Consumer;

  constructor(serviceId: string) {
    super();
    this.serviceId = serviceId;
  }

  getMeta() {
    return {
      serviceType: this.serviceType,
      serviceId: this.serviceId,
      serviceDirection: this.serviceDirection,
    };
  }

  setupConnection(connection: PeerConnection, serviceConfig: ServiceConfiguration): void {
    checkConfig(serviceConfig);
    const channel = new DataChannel();
    channel.ondata = this.handleData.bind(this);
    if (connection.tiebreaker) {
      connection.transmit(serviceConfig, 'data', channel);
    } else {
      connection.receive(serviceConfig, 'data', channel);
    }
  }

  handleData(data: string | Blob | ArrayBuffer | ArrayBufferView) {
    if (typeof data === 'string') {
      const msg = JSON.parse(data);
      this.emit('message', { message_type: msg.messageType, message: msg.message });
    }
  }
}
